### PR TITLE
[MM-18538] Add ability to name a subscription.

### PR DESCRIPTION
**Pull request by jfrerich**
## This PR adds the ability to name a subscription in the following ways:

* Add ability to name a subscription
  * If subscription has a saved name, show it
  * if subscription doesn't have saved name, show "(no name)"
* After click edit on a subscription from the subscription list:
  * if subscription has a name
    * prefill Name with existing name
  * if subscription was created earlier (without a name)
    * leave input null
* Validation
  * will require user to add a name before they can save subscription

## Added to the scope of this ticket
  * Restrictions on creating a subscription name that already exists
    * When editing an existing and creating a new subscription
	* Note the error message added in the image below

image

## Ticket
https://mattermost.atlassian.net/browse/MM-18538

### CODE REVIEW

**Comment by jfrerich** [comment]
@mickmister yes, we can add, and need to, add the check for unique subscription name.  I mentioned in the PR text that this might be placed outside the scope of this ticket, but I suppose it is fine to add it here.

Let's clarify here and I'll add the code in a new commit.  

The channel subscription name is a new field and we want to check that a subscription name is not in use before actually adding the subscription.

In the `Add` method, we will need to check `subscriptionByName` and verify the subscription name does not already exist.
- if name exists, return the error in http.response to the client and inform the user with an error validation message.
- if name does not already exist, continue as usual.

In a nutshell, would you agree?

**Comment by mickmister** [comment]
> I mentioned in the PR text that this might be placed outside the scope of this ticket, but I suppose it is fine to add it here.

Sorry I missed that

@jfrerich Yes your control flow is correct, I think we should also check that the name is present in the request in general as well (even if the client is already doing so).



I think it should only be a conflict if the `name` AND `channel_id` match an existing subscription in my opinion. This would make it so you could have a "Triage Tickets" subscription for the Integrations channel, and a "Triage Tickets" subscription for the Toolkit channel. This is a UX decision. @aaronrothschild, @asaadmahmood What are your thoughts on this?

Also, the uniqueness check should probably run before the call to `ChannelSubscriptions.add()`

This is because of the way `editChannelSubscription` works. It removes the existing sub then adds the new version. The uniqueness check would need to be before the remove happens here:

https://github.com/mattermost/mattermost-plugin-jira/blob/63fd398a9026cdcb825a3b23a246c6163c33e1f4/server/subscribe.go#L286-L291

It also needs to occur in `addChannelSubscription` before the add.

**Comment by jfrerich** [comment]
@mickmister, I agree that the conflict should only occur when a subscription `name` and `channel_id` match an existing subscription.  

Regarding the uniqueness check, I also noticed the difference in `editChannelSubscription` and `addChannelSubscription`.  I'll add new a uniqueness method and check before the remove in `editChannelSubscription`

File: server/subscribe.go [hunk 1]
```diff
@@ -269,6 +275,20 @@ func (p *Plugin) addChannelSubscription(newSubscription *ChannelSubscription) er
 	})
 }
 
+func (p *Plugin) isChannelSubscriptionNameUnique(channelId string, subscription *ChannelSubscription) error {
```
Comment (mickmister) [inline server/subscribe.go]: Suggest changing name to `checkChannelSubscriptionNameUnique` since we are returning an error, rather than a boolean.

File: server/subscribe.go [hunk 2]
```diff
@@ -269,6 +275,20 @@ func (p *Plugin) addChannelSubscription(newSubscription *ChannelSubscription) er
...
+	subs, err := p.getSubscriptionsForChannel(channelId)
+	if err != nil {
+		return err
+	}
+
+	for sub := range subs {
+		if subs[sub].Name == subscription.Name && subs[sub].Id != subscription.Id {
+			return errors.New("Subscription name, \"" + subs[sub].Name + "\", already exists. Please choose another name.")
```
Comment (mickmister) [inline server/subscribe.go]: Nit: Change local var to `subID` and:
```suggestion
			return fmt.Errorf("Subscription name '%s' already exists. Please choose another name.", subs[subID].Name)
```

**Comment by jfrerich** [comment]
@mickister, you're specific changes are in commit https://github.com/mattermost/mattermost-plugin-jira/pull/348/commits/5dc47cae8d9d445d9d99b7c0043a513ce8433e58

Note that I also merged with latest `feature/channel-subscriptions` branch and fixed a bunch of `make check-style` errors.

**Comment by mickmister** [comment]
Great job @jfrerich :+1:

**Comment by aaronrothschild** [comment]
> I think it should only be a conflict if the `name` AND `channel_id` match an existing subscription in my opinion. This would make it so you could have a "Triage Tickets" subscription for the Integrations channel, and a "Triage Tickets" subscription for the Toolkit channel. This is a UX decision. @aaronrothschild, @asaadmahmood What are your thoughts on this?

@jfrerich @mickmister I totally agree with your thinking on this - the user wouldn't be aware of the other subscription names in other channels. They would be confused if we returned an error saying that name was already in use, particularly when they don't have access to check another channel's subscriptions.
